tests/upload_test.php

```diff
@@ -957,4 +958,209 @@ public function test_upload_attendance_for_all_session_statuses(int $timestart,
+            $sessiondates = ['timestart' => time(), 'timefinish' => time()];
+            $datetimeknown = false;
+        }
+
+        $generator = $this->getDataGenerator()->get_plugin_generator('mod_facetoface');
+        $course = $this->getDataGenerator()->create_course();
+        $facetoface = $generator->create_instance(['course' => $course->id]);
+        $student = $this->getDataGenerator()->create_and_enrol($course, 'student');
+        $record = [
+            'facetoface' => $facetoface->id,
+            'capacity' => '3',
+            'allowoverbook' => '0',
+            'details' => 'xyz',
+            'duration' => '1.5',
+            'normalcost' => '111',
+            'discountcost' => '11',
+            'allowcancellations' => '0',
+            'datetimeknown' => $datetimeknown,
+            'sessiondates' => [$sessiondates],
+        ];
+        $session = $generator->create_session($record);
+
+        // Replace contents placeholders.
+        $contents = str_replace(':email', $student->email, $contents);
+        $contents = str_replace(':session', $session->id, $contents);
+
+        // Write csv to stored_file
+        // booking_manager expects it to be in the users draft files.
+        $fs = new file_storage();
+        $storedfile = $fs->create_file_from_string([
+            'contextid' => \context_user::instance($USER->id)->id,
+            'component' => 'user',
+            'filearea' => 'draft',
+            'itemid' => 0,
+            'filepath' => '/',
+            'filename' => uniqid() . '.csv',
+        ], $contents);
+        $bm = new booking_manager($facetoface->id);
+        $bm->load_from_file($storedfile->get_itemid());
+        return [$bm, $session, $student];
+    }
+
+    /**
+     * Test booking_manager validate function
+     *
+     * @param string $contents csv contents
+     * @param string|null $expectedexceptionmessage if given, expect moodle_exception with this message.
+     *
+     * @dataProvider csv_validation_provider
+     */
+    public function test_csv_validation(string $contents, ?string $expectedexceptionmessage = null): void {
+        [$bm, $session, $user] = $this->setup_csv_test($contents);
+        if (!empty($expectedexceptionmessage)) {
+            $this->expectException(\moodle_exception::class);
+            $this->expectExceptionMessage($expectedexceptionmessage);
+        }
+        $bm->validate();
+    }
+
+    /**
+     * Provides CSV processing test cases
+     * @return array
+     */
+    public static function csv_process_provider(): array {
+        // Build a matrix of status, date (past present future and unknown).
+        $dateoptions = [
+            'past' => ['timestart' => time() - DAYSECS * 2, 'timefinish' => time() - DAYSECS],
+            'present' => ['timestart' => time() - DAYSECS, 'timefinish' => time() + DAYSECS],
+            'future' => ['timestart' => time() + DAYSECS, 'timefinish' => time() - DAYSECS * 2],
```

Comment:
```suggestion
            'future' => ['timestart' => time() + DAYSECS, 'timefinish' => time() + DAYSECS * 2],
```
Should timefinish be after timestart?